runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp

```diff
@@ -0,0 +1,268 @@
+          "Remote rest platform: applying IR passes failed.");
+
+    cudaq::info("- Pass manager was applied.");
+
+    opts.llvmModuleBuilder =
+        [](Operation *module,
+           llvm::LLVMContext &llvmContext) -> std::unique_ptr<llvm::Module> {
+      llvmContext.setOpaquePointers(false);
+      auto llvmModule = translateModuleToLLVMIR(module, llvmContext);
+      if (!llvmModule) {
+        llvm::errs() << "Failed to emit LLVM IR\n";
+        return nullptr;
+      }
+      ExecutionEngine::setupTargetTriple(llvmModule.get());
+      return llvmModule;
+    };
+
+    cudaq::info(" - Creating the MLIR ExecutionEngine");
+    auto uniqueJit = llvm::cantFail(ExecutionEngine::create(module, opts));
+    cudaq::info("- JIT Engine created successfully.");
+    return uniqueJit;
+  }
+
+  void invokeMlirKernel(std::unique_ptr<MLIRContext> &contextPtr,
+                        std::string_view irString,
+                        const std::vector<std::string> &passes,
+                        const std::string &entryPointFn) {
+    llvm::SourceMgr sourceMgr;
+    sourceMgr.AddNewSourceBuffer(llvm::MemoryBuffer::getMemBufferCopy(irString),
+                                 llvm::SMLoc());
+    auto module = parseSourceFile<ModuleOp>(sourceMgr, contextPtr.get());
+    auto engine = jitMlirCode(*module, passes);
+    const std::string entryPointFunc =
+        std::string(cudaq::runtime::cudaqGenPrefixName) + entryPointFn;
+    auto fnPtr = llvm::cantFail(engine->lookup(entryPointFunc));
+    if (!fnPtr)
+      throw std::runtime_error("Failed to get entry function");
+
+    auto fn = reinterpret_cast<void (*)()>(fnPtr);
+    // Invoke the kernel
+    fn();
+  }
+
+  void *loadNvqirSimLib(const std::string &simulatorName) {
+    const std::filesystem::path cudaqLibPath{cudaq::getCUDAQLibraryPath()};
+#if defined(__APPLE__) && defined(__MACH__)
+    const auto libSuffix = "dylib";
+#else
+    const auto libSuffix = "so";
+#endif
+    const auto simLibPath =
+        cudaqLibPath.parent_path() /
+        fmt::format("libnvqir-{}.{}", simulatorName, libSuffix);
+    cudaq::info("Request simulator {} at {}", simulatorName,
+                simLibPath.c_str());
+    void *simLibHandle = dlopen(simLibPath.c_str(), RTLD_GLOBAL | RTLD_NOW);
+    if (!simLibHandle) {
+      char *error_msg = dlerror();
+      throw std::runtime_error(fmt::format(
+          "Failed to open simulator backend library: {}.",
+          error_msg ? std::string(error_msg) : std::string("Unknown error")));
+    }
+    auto *sim = cudaq::getUniquePluginInstance<nvqir::CircuitSimulator>(
+        std::string("getCircuitSimulator"), simLibPath.c_str());
+    __nvqir__setCircuitSimulator(sim);
+
+    return simLibHandle;
+  }
+
+  json processRequest(const std::string &reqBody) {
+    static std::size_t g_requestCounter = 0;
+    auto requestJson = json::parse(reqBody);
+    cudaq::RestRequest request(requestJson);
+    const auto reqId = g_requestCounter++;
```

Comment:
Just curious: do we expect multiple threads to be calling this function? If so, do we need a lock?

Reply:
No, we don't expect multi-threading here. I've added some comments in the code to note this (https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/f531656e608fbd1ad84776d8ce1660c4b8a1b46e)